presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java

```diff
@@ -134,94 +194,377 @@ public boolean needsInput()
+                // Spill state didn't change, so process as usual.
+                processProbe(lookupSourceLease.getLookupSource());
+                return Optional.empty();
+            }
 
-        return lookupSourceProvider.withLease(lookupSourceLease -> {
-            return getOutput(lookupSourceLease.getLookupSource());
+            return Optional.of(SpillInfoSnapshot.from(lookupSourceLease));
         });
+
+        if (!spillInfoSnapshotIfSpillChanged.isPresent()) {
+            return;
+        }
+        SpillInfoSnapshot spillInfoSnapshot = spillInfoSnapshotIfSpillChanged.get();
+        long joinPositionWithinPartition;
+        if (joinPosition >= 0) {
+            joinPositionWithinPartition = lookupSourceProvider.withLease(lookupSourceLease -> lookupSourceLease.getLookupSource().joinPositionWithinPartition(joinPosition));
+        }
+        else {
+            joinPositionWithinPartition = -1;
+        }
+
+        /*
+         * Spill state changed. All probe rows that were not processed yet should be treated as regular input (and be partially spilled).
+         * If current row maps to the now-spilled partition, it needs to be saved for later. If it maps to a partition still in memory, it
+         * should be added together with not-yet-processed rows. In either case we need to resume processing the row starting at its
+         * current position in the lookup source.
+         */
+        verify(spillInfoSnapshot.hasSpilled);
+        verify(spillInfoSnapshot.spillEpoch > inputPageSpillEpoch);
+
+        Page currentPage = probe.getPage();
+        int currentPosition = probe.getPosition();
+        long currentJoinPosition = this.joinPosition;
+        boolean currentProbePositionProducedRow = this.currentProbePositionProducedRow;
+        int joinSourcePositions = this.joinSourcePositions;
+
+        probe = null;
+
+        if (currentPosition < 0) {
+            // Processing of the page hasn't been started yet.
+            addInput(currentPage, spillInfoSnapshot);
+        }
+        else {
+            int currentRowPartition = getPartitionGenerator().getPartition(currentPage, currentPosition);
+            boolean currentRowSpilled = spillInfoSnapshot.spillMask.test(currentRowPartition);
+
+            if (currentRowSpilled) {
+                savedRows.merge(
+                        currentRowPartition,
+                        new SavedRow(currentPage, currentPosition, joinPositionWithinPartition, currentProbePositionProducedRow, joinSourcePositions),
+                        (oldValue, newValue) -> {
+                            throw new IllegalStateException(format("How on earth this could happen that partition %s is spilled in the middle of processing twice?", currentRowPartition));
+                        });
+                Page unprocessed = pageTail(currentPage, currentPosition + 1);
+                addInput(unprocessed, spillInfoSnapshot);
```

Comment:
why don't we need to clear this.joinPosition, this.currentProbePositionProducedRow, this.joinSourcePositions... just seems weird that one code path sets them and the other doesn't

Reply:
because `addInput` does that to `joinPosition`, `currentProbePositionProducedRow` is irrelevant (initially there is no current row) and `joinSourcePositions` should be zero (i added a fixup clearing it when we save row)